src/Symfony/Component/String/AbstractString.php

```diff
@@ -0,0 +1,555 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\String;
+
+use Symfony\Component\String\Exception\ExceptionInterface;
+use Symfony\Component\String\Exception\InvalidArgumentException;
+
+/**
+ * Represents a string.
+ *
+ * @author Nicolas Grekas <p@tchwork.com>
+ * @author Hugo Hamon <hugohamon@neuf.fr>
+ *
+ * @throws ExceptionInterface
+ */
+abstract class AbstractString implements \JsonSerializable
+{
+    protected $string = '';
+    protected $locale = null;
+    protected $ignoreCase = false;
+
+    abstract public function __construct(string $string = '', string $locale = null);
+
+    /**
+     * @return static
+     */
+    abstract public function after(string $needle, bool $includeNeedle = false): self;
+
+    /**
+     * @return static
+     */
+    abstract public function afterLast(string $needle, bool $includeNeedle = false): self;
+
+    /**
+     * @return static
+     */
+    abstract public function append(string ...$suffix): self;
+
+    /**
+     * @return static
+     */
+    abstract public function before(string $needle, bool $includeNeedle = false): self;
+
+    /**
+     * @return static
+     */
+    abstract public function beforeLast(string $needle, bool $includeNeedle = false): self;
+
+    /**
+     * @return static[]
+     */
+    abstract public function chunk(int $length = 1): array;
+
+    /**
+     * @return static
+     */
+    public function collapseWhitespace(): self
```

Comment:
Is there such a thing as `final` annotation for public methods and if yes, should it be used on many methods in this class?